src/mgr/QueryParams.cc

```diff
@@ -68,60 +69,61 @@ Mgr::QueryParams::find(const String& name) const
+/// Value can be a comma separated list of integers, or an opaque string.
+Mgr::QueryParam::Pointer
+ParseParamValue(SBuf &valueStr)
 {
-    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
-
-    Parser::Tokenizer tok(paramStr);
-
-    SBuf name;
-    if (!tok.prefix(name, nameChars))
-        return false;
-
-    if (!tok.skip('='))
-        return false; // TODO: support name-only parameters
-
-    static const CharacterSet badValueChars = CharacterSet("param-value-bad", "&= ");
-    SBuf valueStr = tok.remaining();
-    if (valueStr.findFirstOf(badValueChars) != SBuf::npos)
-        return false;
-
-    param.first = SBufToString(name);
+    static const CharacterSet comma("comma",",");
 
+    Parser::Tokenizer tok(valueStr);
     std::vector<int> array;
     int64_t intVal = 0;
     while (tok.int64(intVal, 10, false)) {
-        static const CharacterSet comma("comma",",");
         (void)tok.skipAll(comma);
         Must(intVal >= std::numeric_limits<int>::min());
         Must(intVal <= std::numeric_limits<int>::max());
-        array.emplace_back(int(intVal));
+        array.emplace_back(intVal);
     }
 
     if (tok.atEnd())
-        param.second = new IntParam(array);
+        return new Mgr::IntParam(array);
     else
-        param.second = new StringParam(SBufToString(valueStr));
-
-    return true;
+        return new Mgr::StringParam(SBufToString(valueStr));
 }
 
-bool
+/**
+ * Syntax:
+ *   query  = param *( '&' param ) [ '#' fragment ]
+ *   param  = name '=' value
+ *   name   = [a-zA-Z0-9]+
+ *   value  = *pchar | ( 1*DIGIT *( ',' 1*DIGIT ) )
+ */
+void
 Mgr::QueryParams::Parse(Parser::Tokenizer &tok, QueryParams &aParams)
 {
-    static const CharacterSet paramDelim("query-param","&");
-
-    SBuf foundStr;
-    while (tok.token(foundStr, paramDelim)) {
-        Param param;
-        if (!ParseParam(foundStr, param))
-            return false;
-        aParams.params.push_back(param);
-    }
+    static const CharacterSet nameChars = CharacterSet("param-name", "_") + CharacterSet::ALPHA + CharacterSet::DIGIT;
+    static const CharacterSet valueChars = CharacterSet("param-value", "&= #").complement();
+
+    while (!tok.atEnd() && !tok.skip('#')) {
 
-    // success == no garbage in query-string
-    return tok.atEnd();
+        SBuf nameStr;
+        if (!tok.prefix(nameStr, nameChars))
+            throw TexcHere("invalid character in action name");
+        if (!tok.skip('='))
+            throw TexcHere("missing parameter value");
+
+        SBuf valueStr;
+        if (!tok.prefix(valueStr, valueChars))
+            throw TexcHere("invalid character in parameter value");
+
+        if (tok.skip('=') || tok.skip(' '))
+            throw TexcHere("invalid character in parameter value");
+
+        const auto name = SBufToString(nameStr);
+        const auto value = ParseParamValue(valueStr);
+        aParams.params.emplace_back(name, value);
+    }
```

Comment:
AFAICT, the code does not consume `&` between parameters. If I am right, then there is something wrong with the tests used during this PR work -- they should have revealed this basic parsing problem. Roughly speaking:

- If we do not allow empty parameters, then exactly one `&` should be skipped at the end of every the loop iteration (if it is present).
- Otherwise, all `&` should be consumed in the beginning of every loop iteration (if they are present).

Reply:
Dang. Lost it in the re-write. Fixed now I think. Will re-test in a day or so.

Reply:
Done in ca0de8c